Suspend/Resume and S0ix implementation

ranj063 opened the pull request:
This patchset deals with implementation the suspend/resume flow for PM in the SOF driver. 
The main differences from the previous pull request are as follows:
1. Fixed a bunch of mem leaks
2. stick to attaching the PM callbacks for the pci/acpi/spi device as originally intended and allow
runtime_pm for these devices.
3. Modify load_firmware() to indicate first boot

Still only tested on up squared board with a single pipeline. Multiple pipelines testing is WIP.

ranj063 commented:
@plbossart @lgirdwood This is the new series for suspend/resume implementation. Could you please help review?

ranj063 commented:
@plbossart @lgirdwood Update for 07/25:

I have tested this PR with a 2 pipeline topology on the up squared board containing a playback pipeline and a dmic capture pipeline and it worked without any issues.

lgirdwood commented on sound/soc/sof/topology.c:
```diff
@@ -877,32 +877,47 @@ static int sof_widget_load_buffer(struct snd_soc_component *scomp, int index,
+	ret = sof_ipc_tx_message(sdev->ipc,
+				 buffer->comp.hdr.cmd, buffer, sizeof(*buffer),
+				 r, sizeof(*r));
+
```
newline can be removed here and maybe other places.

lgirdwood commented on sound/soc/sof/topology.c:
```diff
@@ -1019,30 +1057,42 @@ static int sof_widget_load_mixer(struct snd_soc_component *scomp, int index,
+	ret = sof_ipc_tx_message(sdev->ipc,
+				 mixer->comp.hdr.cmd, mixer, sizeof(*mixer), r,
+				 sizeof(*r));
+
```
dito, newline between set and check ret.

lgirdwood commented on sound/soc/sof/topology.c:
```diff
@@ -1083,45 +1138,54 @@ static int sof_widget_load_pga(struct snd_soc_component *scomp, int index,
+	ret = sof_ipc_tx_message(sdev->ipc,
+				 volume->comp.hdr.cmd, volume, sizeof(*volume),
+				 r, sizeof(*r));
+
```
ditto

lgirdwood commented on sound/soc/sof/topology.c:
```diff
@@ -1135,42 +1199,54 @@ static int sof_widget_load_src(struct snd_soc_component *scomp, int index,
+	swidget->private = (void *)src;
+
+	ret = sof_ipc_tx_message(sdev->ipc,
+				 src->comp.hdr.cmd, src, sizeof(*src), r,
```
I've noticed a lot of these where params are spread over three lines where they should fit int two. It makes it easier for me to read if they are on two (and closer to the conditional check).

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -877,32 +877,47 @@ static int sof_widget_load_buffer(struct snd_soc_component *scomp, int index,
 {
```
the second patch does not compile, please make sure each patch can be compiled to avoid breaking git bisect
/data/pbossart/ktest/sof-dev/sound/soc/sof/topology.c: In function ‘snd_sof_free_topology’:
/data/pbossart/ktest/sof-dev/sound/soc/sof/topology.c:2130:48: error: ‘struct snd_sof_dev’ has no member named ‘route_list’; did you mean ‘pcm_list’?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1886,15 +1886,28 @@ static int sof_route_load(struct snd_soc_component *scomp, int index,
 			  struct snd_soc_dapm_route *route)
```
/data/pbossart/ktest/sof-dev/sound/soc/sof/topology.c: In function ‘sof_widget_ready’:
/data/pbossart/ktest/sof-dev/sound/soc/sof/topology.c:1421:5: warning: ‘ret’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  if (ret < 0 || reply.rhdr.error < 0) {

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -9,35 +9,316 @@
  */
 
```
/data/pbossart/ktest/sof-dev/sound/soc/sof/pm.c: In function ‘sof_suspend_streams’:
/data/pbossart/ktest/sof-dev/sound/soc/sof/pm.c:177:9: error: ‘struct snd_sof_pcm’ has no member named ‘restore_stream’
     spcm->restore_stream[dir] = 1;
         ^~
/data/pbossart/ktest/sof-dev/sound/soc/sof/pm.c:194:9: error: ‘struct snd_sof_pcm’ has no member named ‘restore_stream’
     spcm->restore_stream[dir] = 1;
         ^~

ranj063 commented:
@plbossart let me look into the compilation issue. I might have to re-order the commits.

ranj063 commented:
@plbossart @lgirdwood I have addressed both of your comments in the latest update

ranj063 commented:
@plbossart I have updated this pull request based on the findings on suspend/resume tests on chrome. Could I please request you to review the changes? 

The main changes are: 
1. configure all DAI's with the same name
2. Skip virtual routes from being saved to the route list
3. modified pcm trigger to account for resuming active pcm streams in chrome.

naveen-manohar commented:
@plbossart, @ranj063 I have verified D0ix functionality with the mentioned 15 patches(6b799ab)

lgirdwood commented:
@plbossart I'll merge this in upstream v2 when merged.

plbossart commented:
@lgirdwood I think we should do a partial merge. There is an awful amount of code related to topology handling that could do in the first upstream batch (and would work without suspend/resume). It may be a good thing to split this set in two, one as groundwork and the second as suspend/resume/runtime_pm proper.

plbossart commented:
This patch set has become too big to review in details, so I'll merge it and we'll have to deal with issues in follow-up patches.
